File: C++/Problem004.cpp
#include "pch.h"
#include "Problems.h"

#include <vector>
#include <cmath>
#include <string>
#include <algorithm>

using std::vector;
using std::pow;
using std::string;
using std::to_string;


/*
=================
=   Problem 4   =	Largest palindrome product
==================


A palindromic number reads the same both ways. The largest palindrome made from
the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.
*/

// Finds the largest palindrome made from the product of two n-digit specific
// numbers.Iterate through all possible numbers and save all palindroms in a list.
// !!appending to list is costly !!

// @param integer - number of digits for multiplied numbers
// @return integer - largest palindrome of the product of two n digit numbers
unsigned int problem4(unsigned int numDigits) {
	unsigned int minNum = pow(10, numDigits - 1);
	unsigned int maxNum = pow(10, numDigits);
	unsigned int isPrime = 2;
	vector<int> palindroms;
	for (size_t i = minNum; i < maxNum; ++i) {
		for (size_t x = minNum; x < maxNum; ++x) {
			string productString = to_string(i*x);

			// create new variable for reverse string
			string revString{ productString };
			std::reverse(revString.begin(), revString.end());

			if (productString == revString) {
				palindroms.push_back(i*x);
			}
		}
	}
	// !! max_element returns an iterator so need to use * in front !!
	auto maxPalindrome = *max_element(std::begin(palindroms), std::end(palindroms));
	return maxPalindrome;
}

File: Python/problem004.py
# -*- coding: utf-8 -*-
"""
Created on Mon Nov 26 19:17:07 2018

@author: MortZ

Project Euler Problem 4
"""


'''
A palindromic number reads the same both ways. The largest palindrome made from 
the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers.
'''

from timeit import default_timer as timer


'''
Finds the largest palindrome made from the product of two n-digit specific
numbers. Iterate through all possible numbers and save all palindroms in a list. 
!! appending to list is costly !!

@param integer - number of digits for multiplied numbers
@return integer - largest palindrome of the product of two n digit numbers
'''
def maxPalindrom(n):
    assert n >= 1, "A positive integer bigger than 0 should be passed!"
    try:
        minNum = 10**(n-1)
        maxNum = 10**(n) 
        palindroms = []
        for i in range(minNum, maxNum):
            for x in range (minNum, maxNum):
                if (str(i*x) == str(i*x)[::-1]):    # [::-1] reverses a string
                    palindroms.append(i*x)
        return max(palindroms)
    except (TypeError):
        raise TypeError ("A positive integer bigger than 1 should be passed")

def maxPalindrom2(n):
    assert n >= 1, "A positive integer bigger than 0 should be passed!"
    try:
        minNum = 10**(n-1)
        maxNum = 10**(n) 
        palindrom = max(i*x
                        for i in range(minNum, maxNum)
                        for x in range (minNum, maxNum)
                        if (str(i*x) == str(i*x)[::-1])
                        )
        return palindrom
    except (TypeError):
        raise TypeError ("A positive integer bigger than 1 should be passed")


if __name__ == "__main__":
    start = timer()
    numDigits = 3
    print(maxPalindrom2(numDigits))
    end = timer()
    print(end - start)
